### PR TITLE
Refactored Vispy2D renderer under one module 

### PR DESCRIPTION
**Pull request by tushar5526**
`Userspace.py` now only imports and uses vispy as renderer when selected.
`Base.py` and `renderer2D.py` is refactored.

Before working on integrating Cairo and refactoring code, should we work on writing **visual tests** for p5py?
As sooner or later we will need that and manually checking for bugs when we will start integrating Cairo will become hectic.
 
**This PR is not tested robustly, I just tested it on some 2D and 3D examples, and things were working fine :)**

### CODE REVIEW

File: p5/sketch/userspace.py [hunk 1]
```diff
@@ -143,34 +139,35 @@ def run(sketch_setup=None, sketch_draw=None,
...
-    builtins.pixel_x_density = physical_width // width
-    builtins.pixel_y_density = physical_height // height
-    builtins.start_time = time.perf_counter()
+        p5.sketch.timer.start()
 
-    p5.sketch.timer.start()
-
-    app.run()
-    exit()
+        app.run()
```
Comment (parsoyaarihant) [inline p5/sketch/userspace.py:167]: This line is visy specific. I believe if we use another renderer, it will throw an error.

**Comment by parsoyaarihant** [comment]
To test the changes, we can directly checkout your branch locally and test it. We dont have to create a new branch in p5py.

**Comment by parsoyaarihant** [comment]
Nice work! I will test this properly in the coming days. 

For #214 we now need two things:

1.  Refactor the PShape class into the newly created vispy renderer. Basically move the shape.py file into the Vispy2DRenderer folder.
2. Create new methods in Vispy2DRenderer class for the primitive APIs (like line, rectangle, etc). Then call these functions from the primitives.py. The primitive shape functions should not return a PShape object like the currently do, instead they should call the corresponding function in the Vispy2DRenderer.

**Comment by parsoyaarihant** [comment]
Can you put `shaders2d.py` (and shader files) and `openglrenderer` inside the `Vispy2DRenderer`?

File: p5/sketch/userspace.py (hunk 1 above)
Comment (tushar5526) [reply p5/sketch/userspace.py:167]: `app.run()` will be executed only when the renderer is set to `vispy`. 
`app.run()` and all other `vispy` specific calls is under the `if renderer == "vispy"` clause. I think Github's UI is making it confusing to figure out the indents.

Or is it something else that I am missing, can you confirm this @parsoyaarihant ?
Thank you for the review :)

**Comment by tushar5526** [comment]
> 1. Refactor the PShape class into the newly created vispy renderer. Basically move the shape.py file into the Vispy2DRenderer folder.
> 2. Create new methods in Vispy2DRenderer class for the primitive APIs (like line, rectangle, etc). Then call these functions from the primitives.py. The primitive shape functions should not return a PShape object like the currently do, instead they should call the corresponding function in the Vispy2DRenderer.

Thanks for the detailed description of what needs to be done. I will start working on it, but due to exams will be little unactive till the month's end.

**Comment by tushar5526** [comment]
Now, `renderer2d` and `renderer3d` is imported after we check which *renderer mode and renderer backend* is specified by the user.
I have also encapsulated all 3d renderer code ( shader files, renderer3d.py) in Vispy3DRenderer, otherwise `renderer3d` was loading **shader files** before `builtins.current_renderer` was defined.
I tried to generalize things as much as possible so that it's easy to add other backends for both 2D and 3D in the future.

**Comment by tushar5526** [comment]
I will be making PRs to this branch from now onwards or there will be too much code to review :)

File: .gitignore [hunk 2]
```diff
@@ -111,3 +111,8 @@ ENV/
 
 # Profiling results
 profiling/*.prof
+
+.idea/
+test.py
```
Comment (parsoyaarihant) [inline .gitignore:116]: This is useful for development. But we need to discuss and document this properly.

File: p5/core/primitives.py [hunk 3]
```diff
@@ -25,7 +25,7 @@
 from ..pmath import curves
 from ..pmath.utils import SINCOS
 
-from .shape import PShape
+from p5.sketch.Vispy2DRenderer.shape import PShape
```
Comment (parsoyaarihant) [inline p5/core/primitives.py:28]: We dont have to use the entire namespace. We can use `from ..sketch.Vispy2DRenderer.shape import PShape`

Comment (tushar5526) [reply p5/core/primitives.py:28]: These changes in namespaces were done automatically by **pycharm** when I moved (refactored) the files. Should I undo this manually ?

File: .gitignore (hunk 2 above)
Comment (tushar5526) [reply .gitignore]: I have documented this.

**Comment by tushar5526** [comment]
@parsoyaarihant before I redo the APIs, can you confirm is this the intended design we are looking for? 
```
def quad(*args):
    """doc string"""
    
    (parse args here)
    if len(args) == 8:
        p1, p2, p3, p4 = args[:2], args[2:4], args[4:6], args[6:]
    elif len(args) == 4:
        p1, p2, p3, p4 = args
    else:
        raise ValueError("Unexpected number of arguments passed to quad()")
    
    (pass onto renderer then)
    p5.renderer.quad(p1, p2, p3, p4)
```
**Quad method in VispyRenderer Class**

```
    def quad(self, *args):
        p1, p2, p3, p4 = args
        path = [
            Point(*p1),
            Point(*p2),
            Point(*p3),
            Point(*p4)
        ]
        self.render(PShape(vertices=path, shape_type=SType.QUADS))
```

**Comment by ziyaointl** [comment]
Great work! Now that the shaders have moved, you may want to modify MANIFEST.in to include the new shader location.

**Comment by parsoyaarihant** [comment]
> @parsoyaarihant before I redo the APIs, can you confirm is this the intended design we are looking for?

Yes, this is the design we are looking for

File: p5/core/primitives.py (hunk 3 above)
Comment (parsoyaarihant) [reply p5/core/primitives.py:28]: Fix these imports. Then I will merge this PR

**Comment by parsoyaarihant** [comment]
> Before working on integrating Cairo and refactoring code, should we work on writing visual tests for p5py?

Cairo has anti-aliasing and Vispy does not. If we write the tests for Vispy, then they will fail with Cairo.

**Comment by tushar5526** [comment]
I have compiled all the examples present at p5's read the docs for testing purposes. In case you need it, they are under **check_tests** at this [branch](https://github.com/tushar5526/p5/tree/p5Examples)

**Pull request merged**